app/forms/firewall-rules-common.tsx

```diff
@@ -6,115 +6,324 @@
+              onSubmitTextField(e)
+            }
+          }}
+          // TODO: validate here, but it's complicated because it's conditional
+          // on which type is selected
+        />
+      )}
+    </>
+  )
+}
+
+// The "Clear" and "Add …" buttons that appear below the filter input fields
+const ClearAndAddButtons = ({
+  isDirty,
+  onClear,
+  onSubmit,
+  buttonCopy,
+}: {
+  isDirty: boolean
+  onClear: () => void
+  onSubmit: () => void
+  buttonCopy: string
+}) => (
+  <div className="flex justify-end">
+    <Button
+      variant="ghost"
+      size="sm"
+      className="mr-2.5"
+      disabled={!isDirty}
+      onClick={onClear}
+    >
+      Clear
+    </Button>
+    <Button size="sm" onClick={onSubmit}>
+      {buttonCopy}
+    </Button>
+  </div>
+)
+
+type TypeAndValueTableProps = {
+  sectionType: 'target' | 'host'
+  items: ControllerRenderProps<FirewallRuleValues, 'targets' | 'hosts'>
+}
+const TypeAndValueTable = ({ sectionType, items }: TypeAndValueTableProps) => (
+  <MiniTable.Table
+    className="mb-4"
+    aria-label={sectionType === 'target' ? 'Targets' : 'Host filters'}
+  >
+    <MiniTable.Header>
+      <MiniTable.HeadCell>Type</MiniTable.HeadCell>
+      <MiniTable.HeadCell>Value</MiniTable.HeadCell>
+      {/* For remove button */}
+      <MiniTable.HeadCell className="w-12" />
+    </MiniTable.Header>
+    <MiniTable.Body>
+      {items.value.map(({ type, value }, index) => (
+        <MiniTable.Row
+          tabIndex={0}
+          aria-rowindex={index + 1}
+          aria-label={`Name: ${value}, Type: ${type}`}
+          key={`${type}|${value}`}
         >
-          API docs
-        </a>{' '}
-        to learn more about firewall rules.
-      </>
-    }
-  />
+          <MiniTable.Cell>
+            <Badge variant="solid">{type}</Badge>
+          </MiniTable.Cell>
+          <MiniTable.Cell>{value}</MiniTable.Cell>
+          <MiniTable.RemoveCell
+            onClick={() =>
+              items.onChange(
+                items.value.filter((i) => !(i.value === value && i.type === type))
+              )
+            }
+            label={`remove ${sectionType} ${value}`}
+          />
+        </MiniTable.Row>
+      ))}
+    </MiniTable.Body>
+  </MiniTable.Table>
 )
 
-export const CommonFields = ({ error, control, nameTaken }: CommonFieldsProps) => {
-  const portRangeForm = useForm({ defaultValues: portRangeDefaultValues })
+// Given an array of committed items (VPCs, Subnets, Instances) and
+// a list of all items, return the items that are available
+const availableItems = (
+  committedItems: Array<VpcFirewallRuleTarget | VpcFirewallRuleHostFilter>,
+  // Items is conditional because VPC Subnet fetching isn't 100% guaranteed
+  items?: Array<Vpc | VpcSubnet | Instance>
+) => {
+  if (!items) return []
+  return items
+    .map((item) => item.name)
+    .filter((name) => !committedItems.map((ci) => ci.value).includes(name))
```

Comment:
Well, that is juicy. Good catch.

Reply:
(For posterity: we determined that the VPC subnets available to select here should be restricted to the current VPC, rather than allowing the user to pick any VPC and then any subnet. I'm also adjusting the filter to consider both the type and value, to make sure we don't prune the dropdown list too aggressively.)